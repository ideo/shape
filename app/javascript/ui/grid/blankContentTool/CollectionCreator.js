import _ from 'lodash'
import PropTypes from 'prop-types'

import { FormButton } from '~/ui/global/styled/buttons'
import { BctTextField } from '~/ui/global/styled/forms'
import PaddedCardCover from '~/ui/grid/covers/PaddedCardCover'
import v, { KEYS } from '~/utils/variables'
import { routingStore } from '~/stores'
import googleTagManager from '~/vendor/googleTagManager'

class CollectionCreator extends React.Component {
  state = {
    inputText: '',
  }

  onInputChange = e => {
    if (e.target.value.length > v.maxTitleLength) return
    this.setState({
      inputText: e.target.value,
    })
  }

  handleKeyDown = e => {
    if (e.keyCode === KEYS.ESC) {
      this.props.closeBlankContentTool()
    }
  }

  afterCreate = card => {
    googleTagManager.push({
      event: 'formSubmission',
      formType: `Create ${this.dbType || 'Collection'}`,
    })

    // if creating a submissionBox we route you to finish setting up the collection
    if (this.props.type === 'submissionBox')
      routingStore.routeTo('collections', card.record.id)
  }

  createCollection = e => {
    e.preventDefault()
    if (!this.state.inputText) return
    const { createCard, type } = this.props

    createCard(
      {
        // `collection` is the collection being created within the card
        collection_attributes: {
          name: this.state.inputText,
          master_template: type === 'template',
          type: this.dbType,
        },
      },
      {
        afterCreate: this.afterCreate,
      }
    )
  }

  get dbType() {
    const { type } = this.props
    let dbType = null

    if (type === 'submissionBox') dbType = 'Collection::SubmissionBox'
    else if (type === 'testCollection') dbType = 'Collection::TestCollection'
    else if (type === 'foamcoreBoard') dbType = 'Collection::Board'

    return dbType
  }

  get typeName() {
    const { type } = this.props
    // e.g. 'submissionBox' -> 'Submission Box'
    return _.startCase(type)
  }

  render() {
    return (
      <PaddedCardCover>
        <form className="form" onSubmit={this.createCollection}>
          <BctTextField
            autoFocus
            data-cy="CollectionCreatorTextField"
            placeholder={`${this.typeName} name`}
            value={this.state.inputText}
            onChange={this.onInputChange}
            onKeyDown={this.handleKeyDown}
          />
          <FormButton
            data-cy="CollectionCreatorFormButton"
            disabled={this.props.loading}
            width={125}
          >
            Add
          </FormButton>
        </form>
      </PaddedCardCover>
    )
  }
}

CollectionCreator.propTypes = {
  loading: PropTypes.bool.isRequired,
  type: PropTypes.oneOf([
    'collection',
    'template',
    'testCollection',
    'submissionBox',
    'foamcoreBoard',
  ]),
  createCard: PropTypes.func.isRequired,
  closeBlankContentTool: PropTypes.func.isRequired,
}
CollectionCreator.defaultProps = {
  type: 'collection',
}

export default CollectionCreator
