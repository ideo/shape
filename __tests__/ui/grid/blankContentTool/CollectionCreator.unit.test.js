import CollectionCreator from '~/ui/grid/blankContentTool/CollectionCreator'
import expectTreeToMatchSnapshot from '#/helpers/expectTreeToMatchSnapshot'

const e = { preventDefault: jest.fn() }
let wrapper, props, component
describe('MovableGridCard', () => {
  beforeEach(() => {
    props = {
      loading: false,
      type: 'collection',
      createCard: jest.fn(),
      closeBlankContentTool: jest.fn(),
    }
    props.createCard.mockClear()
    wrapper = shallow(<CollectionCreator {...props} />)
    component = wrapper.instance()
  })

  it('renders snapshot', () => {
    expectTreeToMatchSnapshot(wrapper)
  })

  it('renders a BctTextField', () => {
    expect(wrapper.find('BctTextField').props().placeholder).toEqual(
      'Collection name'
    )
    expect(wrapper.find('BctTextField').props().autoFocus).toBeTruthy()
  })

  describe('createCollection', () => {
    it('calls createCard with input name', () => {
      component.state = {
        inputText: 'New Projects',
      }
      component.createCollection(e)
      expect(props.createCard).toHaveBeenCalledWith(
        {
          collection_attributes: {
            name: component.state.inputText,
            master_template: false,
            type: null,
          },
        },
        {
          afterCreate: null,
        }
      )
    })

    describe('with SubmissionBox', () => {
      beforeEach(() => {
        props.type = 'submissionBox'
        props.createCard.mockClear()
        wrapper = shallow(<CollectionCreator {...props} />)
        component = wrapper.instance()
      })

      it('calls createCard with input name', () => {
        component.state = {
          inputText: 'Challenge #1',
        }
        component.createCollection(e)
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: false,
              type: 'Collection::SubmissionBox',
            },
          },
          {
            afterCreate: expect.any(Function),
          }
        )
      })
    })

    describe('with TestCollection', () => {
      beforeEach(() => {
        props.type = 'testCollection'
        props.createCard.mockClear()
        wrapper = shallow(<CollectionCreator {...props} />)
        component = wrapper.instance()
      })

      it('calls createCard with input name', () => {
        component.state = {
          inputText: 'My New Test',
        }
        component.createCollection(e)
        expect(props.createCard).toHaveBeenCalledWith(
          {
            collection_attributes: {
              name: component.state.inputText,
              master_template: false,
              type: 'Collection::TestCollection',
            },
          },
          {
            afterCreate: null,
          }
        )
      })
    })
  })
})
