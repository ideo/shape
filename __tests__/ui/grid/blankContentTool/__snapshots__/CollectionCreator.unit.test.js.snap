// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovableGridCard renders snapshot 1`] = `
<styled.div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <BctTextField
      autoFocus={true}
      data-cy="CollectionCreatorTextField"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Collection name"
      value=""
    />
    <StyledFormButton
      color="#120f0e"
      data-cy="CollectionCreatorFormButton"
      disabled={false}
      width={125}
    >
      Add
    </StyledFormButton>
  </form>
</styled.div>
`;
